Updated VRControls and VREffect to support WebVR v1

toji opened the pull request:
Retained support for older versions of the API as well. Class interfaces
are unchanged.

toji committed:
Updated VRControls and VREffect to support WebVR v1

Retained support for older versions of the API as well. Class interfaces
are unchanged.

mrdoob committed:
Fixed jsfiddle. See #8185.

toji commented:
Would like to get this into the main repo in advance of us releasing the new API/builds so that people can pick up the latest Three.js and be ready to go!

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,29 +69,59 @@ THREE.VREffect = function ( renderer, onError ) {
+
+			} else {
+
+				console.error("No compatible requestFullscreen method found.");
```
Mr.doob style: `console.error( '...' ); ` (also no punctuation)

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,29 +69,59 @@ THREE.VREffect = function ( renderer, onError ) {
 
 	}, false );
 
 	this.setFullScreen = function ( boolean ) {
```
perhaps this should be renamed to `requestPresent` or some name that isn't fullscreen specific

toji replied:
I'm fine adding an alias for something like `setPresenting( boolean )`, but my primary goal here was to ensure this file could be update in place with no other code changes required by the developer, so `setFullscreen` needs to stick around for a while longer at least.

cvan replied:
understood, that makes sense. so it can be removed more easily in the future, perhaps create a new method, alias the old one and a deprecation warning. similar thing was done to [`zeroSensor`](https://github.com/toji/three.js/blob/webvr_v1/examples/js/controls/VRControls.js#L121-L126).

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,29 +69,59 @@ THREE.VREffect = function ( renderer, onError ) {
+
+	}, false );
+
+	window.addEventListener( 'vrdisplaypresentchange', function ( event ) {
```
`event` is unused (as well as [above](#diff-9241b49ed6cabf364c42e019568a978eL65) too)

mrdoob commented:
@borismus looks good to you?

toji commented:
Updated the commit to take into account @cvan's comments.

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,72 @@ THREE.VREffect = function ( renderer, onError ) {
+			} else {
 
-			canvas.webkitRequestFullscreen( { vrDisplay: vrHMD } );
+				console.error( 'No compatible requestFullscreen method found.' );
```
nit: remove period à la [this console message](https://github.com/mrdoob/three.js/pull/8221/files#diff-4a931a726c43fb78dbae61c51c5a4ab7R27)

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,72 @@ THREE.VREffect = function ( renderer, onError ) {
+
+			} else {
+
+				vrHMD.exitPresent();
```
interesting that this method handles requesting and exiting with the new API, but only entering FS is handled with the old API

toji replied:
I just noticed that myself. Tempted to dig in and fix it but I'm happy to label it a "pre-existing condition" for the moment. :)

cvan replied:
return this one too

cvan commented on examples/js/controls/VRControls.js:
```diff
@@ -13,7 +13,8 @@ THREE.VRControls = function ( object, onError ) {
 		for ( var i = 0; i < devices.length; i ++ ) {
 
-			if ( devices[ i ] instanceof PositionSensorVRDevice ) {
+			if ( devices[ i ] instanceof VRDisplay ||
```
can you check `'VRDisplay' in window`? this errors in Firefox and your older Chromium WebVR builds

borismus replied:
+1 instanceof arguments needs to be defined.

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -12,17 +12,26 @@
 		for ( var i = 0; i < devices.length; i ++ ) {
 
-			if ( devices[ i ] instanceof HMDVRDevice ) {
+			if ( devices[ i ] instanceof VRDisplay ) {
```
can you check `'VRDisplay' in window`? this errors in Firefox and your older Chromium WebVR builds

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,72 @@ THREE.VREffect = function ( renderer, onError ) {
+
+			if ( boolean ) {
+
+				vrHMD.requestPresent( { source: canvas } );
```
should probably catch and log any errors

cvan replied:
don't want to catch anything?

cvan replied:
actually, could we return this Promise? that'd be useful

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,72 @@ THREE.VREffect = function ( renderer, onError ) {
+			} else {
 
-			canvas.webkitRequestFullscreen( { vrDisplay: vrHMD } );
+				console.error( 'No compatible requestFullscreen method found' );
```
period for consistency

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,72 @@ THREE.VREffect = function ( renderer, onError ) {
+
+	window.addEventListener( 'vrdisplaypresentchange', function () {
+
+		isPresenting = vrHMD.isPresenting;
```
though unlikely, it's possible that `vrHMD` is undefined

cvan commented:
I tested these changes with the [webvr interactive cubes example](http://threejs.org/examples/#webvr_cubes), and the scene doesn't render correctly in the DK2. everything appears ultra zoomed in. dropped `VREffect` into other WebVR scenes and was able to consistently reproduce breakage.

jzitelli commented on examples/webvr_cubes.html:
```diff
@@ -127,7 +127,7 @@
 
```
@toji I added `preserveDrawingBuffer: true` to the WebGLRenderer options (per your example: https://github.com/toji/webvr-samples/blob/gh-pages/04-simple-mirroring.html), it fixed the judder that I experienced on DK2... I guess the `preserveDrawingBuffer` property should be set according to the `hasExternalDisplay` capability in all of the webvr examples?

jzitelli replied:
forget it, I think the judder was really just due to naturally low FPS (raycasting against all those boxes I guess).  If I lower the number of boxes in the examples it is nice and smooth, whether `preserveDrawingBuffer` is set or not.  `webvr_stereo_pano` is also working well without it.

jzitelli commented:
@cvan I reverted to setting the viewport / scissor sizes using the values from `renderer.getSize` (instead of using the values from eye parameters), which fixed a similar issue I was having... maybe the `renderWidth` and `renderHeight` eye parameters are not being set correctly for DK2?

cvan commented:
@jzitelli: thanks!
@toji: check out https://github.com/jzitelli/three.js/compare/dev...jzitelli:webvr_v1 to see if you want to borrow some of those changes.

borismus commented:
LGTM, though I haven't done any manual testing.

toji committed:
Fixed broken VR rendering

When rendering the the recommended render width and height should be
ignored and viewports should be constructed off of the current
backbuffer’s actual width and height.

Also fixed some compatibility issues with Firefox.

toji commented:
Fixed a few issues in this patch, including the render.getSize issue @jzitelli mentioned (though I didn't pull over the rest of his patch. The issue is that the renderWidth/renderHeight is the _recommended_ size in order to get a 1:1 pixel match post-distortion. When actually rendering, however, you want to use whatever the actual backbuffer size is. Confirmed that the latest version of my patch works with all three webvr examples (and works with the webvr-polyfill! Though I didn't commit that.)

cvan commented on README.md:
````diff
@@ -61,7 +61,7 @@ This code creates a scene, a camera, and a geometric cube, and it adds the cube
 </script>
 ```
-If everything went well you should see [this](http://jsfiddle.net/f17Lz5ux/).
+If everything went well you should see [this](http://jsfiddle.net/hfj7gm6t/).
````
do you know why http://threejs.org/build/three.min.js !== http://cdn.rawgit.com/mrdoob/three.js/master/build/three.min.js?

mrdoob replied:
See #8185.

cvan commented on examples/js/controls/VRControls.js:
```diff
@@ -13,7 +13,8 @@ THREE.VRControls = function ( object, onError ) {
 		for ( var i = 0; i < devices.length; i ++ ) {
 
-			if ( devices[ i ] instanceof PositionSensorVRDevice ) {
+			if ( ('VRDisplay' in window && devices[ i ] instanceof VRDisplay) ||
```
I think mr. doob likes more spaces in the parens

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,76 @@ THREE.VREffect = function ( renderer, onError ) {
 	};
 
+	this.requestPresent = function () {
+		this.setFullScreen( true );
```
`return ` so we can get back a Promise

cvan replied:
nit: mr.doob newlines before and after this line

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,76 @@ THREE.VREffect = function ( renderer, onError ) {
+	};
+
+	this.exitPresent = function () {
+		this.setFullScreen( false );
```
`return ` so we can get back a Promise

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -129,28 +198,23 @@ THREE.VREffect = function ( renderer, onError ) {
 
 			renderer.setScissorTest( false );
 
+			if ( isPresenting && !deprecatedAPI ) {
```
isn't [`isPresenting` already guaranteed to be `true`](#diff-9241b49ed6cabf364c42e019568a978eR152)?

cvan commented on examples/webvr_cubes.html:
```diff
@@ -127,7 +127,7 @@
 				var fullScreenButton = document.querySelector( '.full-screen' );
 
-				if ( navigator.getVRDevices === undefined ) {
+				if ( navigator.getVRDisplays === undefined && navigator.getVRDevices === undefined ) {
```
perhaps could introduce a method `THREE.VREffect.hasVRSupport`?

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,76 @@ THREE.VREffect = function ( renderer, onError ) {
+	window.addEventListener( 'vrdisplaypresentchange', function () {
 
-		isFullscreen = document.mozFullScreenElement || document.webkitFullscreenElement;
+		if (vrHMD) {
```
if you want to save a few lines:

```js
isPresenting = vrHMD && vrHMD.isPresenting;
```

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,76 @@ THREE.VREffect = function ( renderer, onError ) {
+
+		if ( deprecatedAPI ) {
+
+			isPresenting = document.mozFullScreenElement || document.webkitFullscreenElement;
```
I know this is how this code worked before, but shouldn't this actually check for `vrHMD`? because otherwise you could fullscreen anything without a `vrDisplay` or HMD and `isPresenting` would be `true`.

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -12,17 +12,26 @@
 
 				vrHMD = devices[ i ];
+				deprecatedAPI = false;
+				break; // We keep the first we encounter
```
it'll be nice when hotswapping is supported in the builds

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,76 @@ THREE.VREffect = function ( renderer, onError ) {
+		}
 
 	}, false );
 
```
I noticed that the three.js examples do not provide a method of exiting VR. so in your latest WebVR 1.0 API Chromium builds, if the developer doesn't explicitly stop presenting, the content stays frozen in the headset (until relaunching the browser - and sometimes I've noticed it takes a system restart).

how about adding this to prevent that footgun?

```js
window.addEventListener( 'beforeunload', this.exitPresent.bind(this) );
```

cvan replied:
perhaps also mapping to Escape keypress?

jzitelli commented on examples/js/controls/VRControls.js:
```diff
@@ -47,17 +53,38 @@ THREE.VRControls = function ( object, onError ) {
 
 			var vrInput = vrInputs[ i ];
```
Should there be iteration through `vrInputs` anymore?  I understand that this is how it's been, but it never made sense to me, wouldn't the update always be based on the last `PositionSensorVRDevice` pushed to the array?  It makes even less sense now that `VRDisplay` handles pose too - in `VREffect` we use the first `VRDisplay`, but `VRControl` will update based on the last `VRDisplay`... I'd suggest dropping the `vrInputs` array and pick the first device in both the new and deprecated code paths.

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,34 +69,76 @@ THREE.VREffect = function ( renderer, onError ) {
-			canvas.webkitRequestFullscreen( { vrDisplay: vrHMD } );
+			} else {
+
+				console.error( 'No compatible requestFullscreen method found.' );
```
I wish this method (always) returned a Promise so we could `reject` here - would be useful for handling the Safari/iOS case, for example

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -93,17 +149,26 @@ THREE.VREffect = function ( renderer, onError ) {
-			renderRectR = eyeParamsR.renderRect;
+			if ( !deprecatedAPI ) {
+
+				eyeTranslationL.fromArray(eyeParamsL.offset);
```
spaces around arg

cvan commented on examples/webvr_stereo_pano.html:
```diff
@@ -59,7 +59,7 @@
 
 
-			if ( navigator.getVRDevices ) {
+			if ( navigator.getVRDisplays || navigator.getVRDevices ) {
```
it looks like there's [a bit more](https://github.com/toji/three.js/blob/webvr_v1/examples/webvr_stereo_pano.html#L214) that has to be done with this example to make it work with the new API.

cvan replied:
@toji: just making sure you saw this

cvan commented:
I was seeing some major jank in A-Frame scenes using this until @ashconnell independently encountered the same issue and arrived at [this patch](https://github.com/toji/three.js/compare/webvr_v1...ashconnell:webvr_v1?expand=1).

And here's [my diff](https://github.com/MozVR/webvr-libs/commit/854f053) for the version of `THREE.VREffect` I'm maintaining using until these changes are stable.

jzitelli commented:
@cvan this is good, less invasive patch to the problem that I detailed in #7176 - I was holding off on mentioning it here since it's sort of an independent issue from updating to the new WebVR spec, but I'd recommend it be merged, either in this PR if people don't mind or one right after this is closed!

toji committed:
More fixes to VRControls/VREffect

Based on suggestions by @cvan and @jzitelli

toji commented:
Addressed most of the feedback in this thread.

@cvan: I'm reluctant to add the beforeunload handler because the browser should have that behavior anyway. The fact that my builds don't right now is a bug that I need to correct ASAP. I'm also reluctant to add an escape key handler or similar simply because that might conflict with application behavior.

cvan commented:
> @cvan: I'm reluctant to add the beforeunload handler because the browser should have that behavior anyway. The fact that my builds don't right now is a bug that I need to correct ASAP. I'm also reluctant to add an escape key handler or similar simply because that might conflict with application behavior.

agreed - per https://github.com/toji/chrome-webvr-issues/issues/48#issuecomment-191094320

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,31 +69,88 @@ THREE.VREffect = function ( renderer, onError ) {
 
-		if ( vrHMD === undefined ) return;
-		if ( isFullscreen === boolean ) return;
+		return new Promise(function(resolve, reject) {
```
`function(` -> `function (`

cvan replied:
I'd prefer to return a Promise, but I noticed that @mrdoob isn't using promises anywhere else in the codebase. if you don't want to polyfill, perhaps you could do some conditional logic that returns a `Promise` only if `Promise` is defined on `window`. and then possibly define a `reject` callback that does `console.error` perhaps?

cvan replied:
whichever you / @mrdoob decide, we should make sure that this returns a Promise because it's very important going forward that we use Promises, as the WebVR API makes significant use of them. and it's very useful/necessary to be able to chain the Promises.

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,31 +69,88 @@ THREE.VREffect = function ( renderer, onError ) {
 
-		if ( canvas.mozRequestFullScreen ) {
+			if ( vrHMD === undefined ) {
+				reject( new Error( 'No VR hardware found.' ) );
```
newlines

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,31 +69,88 @@ THREE.VREffect = function ( renderer, onError ) {
+
+				}
+
+				return;
```
unneeded

cvan commented on examples/js/effects/VREffect.js:
```diff
@@ -55,31 +69,88 @@ THREE.VREffect = function ( renderer, onError ) {
+
+					console.error( 'No compatible requestFullscreen method found.' );
+					reject( new Error( 'No compatible requestFullscreen method found.' ) );
+
```
:+1:

cvan commented:
@toji: really appreciate you making these changes (especially with the perf fixes + Promises)

toji committed:
VREffect style fixes

toji commented:
Fixed the style issues mentioned by @cvan (Sorry, I _really_ struggle with doobstyle for some reason...)

I'll let @mrdoob weigh in on the Promise question before I make any further changes there.

mrdoob commented:
Thanks guys!
